controllers/replication.storage/volumegroupreplication_controller.go

```diff
@@ -33,21 +56,533 @@ type VolumeGroupReplicationReconciler struct {
+		}
+
+		if vgrContentObj.Name != "" {
+			err = r.Client.Delete(ctx, vgrContentObj)
+			if err != nil {
+				logger.Error(err, "failed to delete dependent volumeGroupReplicationContent resource")
+				return reconcile.Result{}, err
+			}
+		}
+
+		if err = RemoveFinalizerFromVGR(r.Client, logger, instance); err != nil {
+			logger.Error(err, "failed to remove VolumeGroupReplication finalizer")
+			return reconcile.Result{}, err
+		}
+
+		return reconcile.Result{}, nil
+	}
+
+	// Update VGR status based on VR Status
+	instance.Status.VolumeReplicationStatus = vrObj.Status
+	err = r.Client.Status().Update(ctx, instance)
+	if err != nil {
+		logger.Error(err, "failed to update volumeGroupReplication instance's status", "VGRName", instance.Name)
+		return reconcile.Result{}, err
+	}
+
+	return reconcile.Result{}, nil
 }
 
 // SetupWithManager sets up the controller with the Manager.
 func (r *VolumeGroupReplicationReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	err := r.waitForGroupCrds()
+	if err != nil {
+		return err
+	}
+
+	// Only reconcile for status update events
+	skipUpdates := predicate.Funcs{
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			return false
+		},
+		DeleteFunc: func(e event.DeleteEvent) bool {
+			return false
+		},
+		CreateFunc: func(e event.CreateEvent) bool {
+			return false
+		},
+		GenericFunc: func(e event.GenericEvent) bool {
+			return false
+		},
+	}
+
+	enqueueVGRRequest := handler.EnqueueRequestsFromMapFunc(
+		func(context context.Context, obj client.Object) []reconcile.Request {
+			// Get the VolumeGroupReplication name,namespace
+			var vgrName, vgrNamespace string
+			objAnnotations := obj.GetAnnotations()
+			for k, v := range objAnnotations {
+				if k == "replication.storage.openshift.io/volumegroupref" {
+					vgrName = strings.Split(v, "/")[0]
+					vgrNamespace = strings.Split(v, "/")[1]
+					break
+				}
+			}
+
+			// Skip reconcile if the triggering resource is not a sub-resource of VGR
+			if vgrName == "" || vgrNamespace == "" {
+				return []reconcile.Request{}
+			}
+
+			// Check if the resource is present in the cluster
+			vgrObj := &replicationv1alpha1.VolumeGroupReplication{}
+			logger := log.FromContext(context)
+			err := r.Client.Get(context, types.NamespacedName{Name: vgrName, Namespace: vgrNamespace}, vgrObj)
+			if err != nil {
+				logger.Error(err, "Unable to get VolumeGroupReplication resource")
+				return []reconcile.Request{}
+			}
+
+			// Return name and namespace of the VoleumGroupReplication resource
+			return []reconcile.Request{
+				{
+					NamespacedName: types.NamespacedName{
+						Namespace: vgrObj.Namespace,
+						Name:      vgrObj.Name,
+					},
+				},
+			}
+		},
+	)
+
+	enqueueVGRForPVCRequest := handler.EnqueueRequestsFromMapFunc(
+		func(context context.Context, obj client.Object) []reconcile.Request {
+			// Check if the PVC has any labels defined
+			objLabels := obj.GetLabels()
+			if len(objLabels) == 0 {
+				return []reconcile.Request{}
```

Comment:
there can be a case where someone removed the label from the PVC, we need to remove that from the VGR as well.

Reply:
From what I understand from the [docs](https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/handler#EnqueueRequestsFromMapFunc), the enqueue function would be ran for both the old,new objects in case of update function. So, when the request is ran for older object, the details from the pvc annotation will be fetched and the VGR reconcile request will be enqueued, and since the object is already updated in the cluster the selector won't inlcude that PVC.
Though this is all theoritical, will test it to confirm.